pkg/shared/util.go

```diff
@@ -121,3 +121,45 @@ func StatefulSetContainers(
 
 	return statefulSet.Spec.Template.Spec.Containers
 }
+
+// GetRoleStatusByName looks for the RoleStatus
+// in KubeDirectorCluster.KubeDirectorClusterStatus.Roles[] array
+// by the passed role name and, if RoleStatus exists, returns its address or returns nil and error
+func GetRoleStatusByName(
+	cr *kdv1.KubeDirectorCluster,
+	roleName string,
+) (*kdv1.RoleStatus, error) {
+
+	for i, r := range cr.Status.Roles {
+		if r.Name == roleName {
+			return &cr.Status.Roles[i], nil
+		}
+	}
+	return nil, fmt.Errorf("RoleStatus for %s role name was not found", roleName)
+}
+
+// RoleStatusIsUpgrading checks, if some cluster member of the
+// passed role is currently in upgrading state
+// If role status is not found by passed roleName returns false
+func RoleStatusIsUpgrading(
+	cr *kdv1.KubeDirectorCluster,
+	roleName string,
+) bool {
+
+	rs, err := GetRoleStatusByName(cr, roleName)
+
+	if err != nil || rs.UpgradingMembers == nil {
+		return false
+	}
+
+	return len(rs.UpgradingMembers) > 0
+}
+
+// ClusterIsReady checks, is cluster state equals to "configured" literal
+func ClusterIsReady(
+	cr *kdv1.KubeDirectorCluster,
+) bool {
+	return cr.Status.State == "configured"
+}
+
+var ClusterRollbackInfoMap = make(map[string]*kdv1.RollbackInfo)
```

Comment:
Relying on an in-memory piece of state is tricky since we will lose it if KD gets restarted. We either have to store all the state we need in the kdcluster status, or we have to be able to reconstruct it when KD restarts. In this case I think we would need to store any necessary stuff in the kdcluster status.

This is related to the "New kdcluster status fields" part of our shared design/discussion doc, where it's proposed that we will need info at the top level of the status stanza about what the previous kdapp was, if the spec kdapp has been changed and we're in the middle of an upgrade or rollback.